### PR TITLE
Added In Memory Tree structure to index ACLs

### PR DESCRIPTION
**Pull request by umbreak**
Fixes BlueBrain/nexus#237

### CODE REVIEW

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  * Every key on the map is a [[Path]] and its values are a set of children [[Path]]s. In this way one can
+  * navigate down the tree.
+  *
+  * Note: This implementation is not Thread safe but can be used safely within an Actor.
+  *
+  * @param tree the data structure used to build the tree with the parent paths and the children paths
+  * @param acls a data structure used to store the ACLs for a path
+  * @tparam F the monadic type
+  */
+class InMemoryAclsTree[F[_]: Applicative] private (tree: mutable.Map[Path, Set[Path]],
```
[the remark this answers is not in the document]
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala]: It's obviously less efficient for writes, but good enough; we can later consider an implementation with ConcurrentHashMaps to see the scores on reads and writes using multiple threads.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+      * If the provided ''key'' is not already associated with a value, associates it with the given value.
+      * Otherwise, replaces the value with the results of the given ''f'' function. This operation is not atomic.
+      *
+      * @param key   the key with which the specified value is to be associated
+      * @param value the value to use if absent
+      * @param f     the function to recompute a value if present
+      */
+    def merge(key: K, value: V, f: V => V): Unit =
```
Comment (wwajerowicz) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala]: This name is quite confusing, if `f` is a function which doesn't merge `value` and `c`.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                                                   acls: mutable.Map[Path, AccessControlList])
+    extends AclsIndex[F] {
+
+  private val any = "*"
+
+  override def replace(path: Path, acl: AccessControlList): F[Unit] = {
+    @tailrec
+    def inner(p: Path, children: Set[Path], aclOpt: Option[AccessControlList]): Unit = {
+      tree.merge(p, children, _ ++ children)
+      aclOpt.map(a => acls += (p -> a))
```
Comment (wwajerowicz) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala]: This line can be moved outside of `inner` to simplify.

File: src/test/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTreeBenchmark.scala [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  * listBigAclOrgs             thrpt   10   27121,272 ±  501,287  ops/s
+  * listBigAclProjectsOnOrg    thrpt   10   52515,036 ± 1729,156  ops/s
+  * listBigAllProjects         thrpt   10   31136,426 ± 3893,653  ops/s
+  * listSmallAclOrgs           thrpt   10   41252,103 ±  690,302  ops/s
+  * listSmallAclProjectsOnOrg  thrpt   10  134938,516 ± 3439,873  ops/s
+  * listSmallAllProjects       thrpt   10   51266,234 ±  945,133  ops/s
+
+  */
+@State(Scope.Thread)
+class InMemoryAclsTreeBenchmark extends Randomness {
```
[the remark this answers is not in the document]
Comment (umbreak) [reply src/test/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTreeBenchmark.scala]: I've previously run some other tests with all combinations of ancestors and self to determine which one is the slowest and it is  `self=false` and `ancestors=true`.

That's why then I run all the tests with those values.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala (hunk 2 above)
Comment (umbreak) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala]: It is the same naming used from the [concurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) and they don't seem to feel is weird :)

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/AccessControlList.scala [hunk 5]
```diff
@@ -5,39 +5,39 @@ import ch.epfl.bluebrain.nexus.commons.types.identity.Identity
...
     */
-  def toMap: Map[Identity, Permissions] = acl.map { case AccessControl(id, ps) => id -> ps }.toMap
+  def permissions: Set[Permission] = value.foldLeft(Set.empty[Permission]) { case (acc, (_, perms)) => acc ++ perms }
 
   /**
-    * @return a collapsed [[Permissions]] from all the identities
+    * @return ''true'' if the underlying list is empty or if any pair is found with an empty permissions set
     */
-  def permissions: Permissions = acl.foldLeft(Permissions.empty)(_ ++ _.permissions)
+  def hasVoidPermissions: Boolean = value.isEmpty || value.exists { case (_, perms) => perms.isEmpty }
```
[the remark this answers is not in the document]
Comment (hygt) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/types/AccessControlList.scala]: This is not quite `isEmpty`, and afaik only used when we check that a patch operation doesn't try to add/remove void permissions.

**Pull request merged**